File: 1329-sort-the-matrix-diagonally/1329-sort-the-matrix-diagonally.go
import "sort"
func diagonalSort(mat [][]int) [][]int {
    for r:=0; r<len(mat); r++ {
        sort.Sort(&diag{mat, r,0})
    }
    for c:=1; c< len(mat[0]); c++{
        sort.Sort(&diag{mat, 0,c})
    }
    return mat
}
type diag struct {
    a [][]int
    r,c int
}

func (d *diag)Less(i,j int)bool {
    return d.a[d.r+i][d.c+i] < d.a[d.r+j][d.c+j]
}

func (d *diag)Swap(i,j int) {
    d.a[d.r+i][d.c+i], d.a[d.r+j][d.c+j] = d.a[d.r+j][d.c+j], d.a[d.r+i][d.c+i]
}

func (d *diag)Len()int{
    dr := len(d.a) - d.r
    dc := len(d.a[0]) - d.c
    if dr<dc {
        return dr
    }
    return dc
}

File: 1329-sort-the-matrix-diagonally/1329-sort-the-matrix-diagonally.py
class Solution:
    def diagonalSort(self, mat: List[List[int]]) -> List[List[int]]:
        n, m = len(mat), len(mat[0])
        diagonals = defaultdict(list) 
        
        for i in range(n):
            for j in range(m):
                diagonals[i-j].append(mat[i][j])
        
        for k in diagonals:
            diagonals[k].sort(reverse = True)
        
        
        for i in range(n):
            for j in range(m):
                mat[i][j] = diagonals[i-j].pop()
        
        return mat